### PR TITLE
Allocate machine with interface and storage constraints

### PR DESCRIPTION
**Pull request by howbazaar**
After talking with the MaaS folks, I got a better understanding of the allocation process.

In particular the mapping of storage devices and interfaces to the constraints requested.
There is a dictionary in the response structure of allocate that isn't a standard machine attribute that indicates how the machine was chosen. The keys in this dictionary are only specified when constraints of that type are specified.

### CODE REVIEW

File: controller.go [hunk 1]
```diff
@@ -286,35 +286,65 @@ type AllocateMachineArgs struct {
...
-	NotNetworks []string
-	Zone        string
-	NotInZone   []string
-	AgentName   string
-	Comment     string
-	DryRun      bool
+	Zone      string
+	NotInZone []string
+	// Storage is an optional value listing one or more storage constraints.
+	// Format is an optional label, followed by an optional colon, then size
```
Comment (mjs) [inline controller.go]: s/Format/The format/

File: controller.go [hunk 2]
```diff
@@ -286,35 +286,65 @@ type AllocateMachineArgs struct {
...
+	//
+	//   200(ssd,removable),400(ssd),300
+	//   - 200GB disk with ssd and removable tag
+	//   - 400GB disk with ssd tag
+	//   - 300GB disk
+	//
+	//  root:80(ssd),data:400
+	//  - 80+GB disk with ssd tag, name the constraint "root"
+	//  - 400+GB disk, name the consrtaint "data"
+	Storage string
```
Comment (mjs) [inline controller.go]: Wouldn't it be cleaner to accept this as a []string or better yet a []StorageSpec, where StorageSpec has fields for the size, labels, and tags. The caller shouldn't have to care about how MAAS wants this to be formatted.

File: controller.go [hunk 3]
```diff
@@ -286,35 +286,65 @@ type AllocateMachineArgs struct {
...
+	// Interfaces is an optional value listing one or more interface constraints.
+	// Format of the string is semi-colon delimited labelled constraints. Each
+	// labelled constraint has the format:
+	//   <label>:<key>=<value>[,<key2>=<value2>[,...]]'
+	// The label is used in the ConstraintMatch structure returned from
+	// AllocateMachine. Valid keys are:
+	//     'space', 'not_space', 'fabric_class', 'not_fabric_class',
+	//     'subnet_cidr', 'not_subnet_cidr', 'vid', 'not_vid',
+	//     'fabric', 'not_fabric', 'subnet', 'not_subnet', 'mode'
+	Interfaces string
```
Comment (mjs) [inline controller.go]: As above

**Comment by mjs** [comment]
LGTM. Looks much better.

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**